lib/iris/analysis/__init__.py

```diff
@@ -3133,3 +3136,192 @@ def regridder(self, src_grid, target_grid):
+
+        if data is not None and hasattr(data, "mask"):
+            if is_masked(data):
+                mask_type = MaskTypes.SOME_TRUE
+            else:
+                if data.mask is ma.nomask:
+                    mask_type = MaskTypes.SCALAR_FALSE
+                else:
+                    mask_type = MaskTypes.ARRAY_FALSE
+        else:
+            mask_type = MaskTypes.NOT_APPLICABLE
+
+        return mask_type
+
+    def lift_mask(array_or_cube, cube_store):
+        # Upstream checks should guarantee this is a masked array.
+        data_masked = get_data(array_or_cube)
+        array_lib = get_array_lib(data_masked)
+        data_lifted = array_lib.ma.getdata(data_masked)
+
+        if is_cube(array_or_cube):
+            cube_store.append(array_or_cube)
+            array_or_cube.data = data_lifted
+        else:
+            array_or_cube = data_lifted
+
+        return array_or_cube
+
+    def re_apply_mask(array_or_cube, mask_type):
+        data = get_data(array_or_cube)
+        array_lib = get_array_lib(data)
+
+        if data is not None and not has_mask(data):
+            if mask_type == MaskTypes.SCALAR_FALSE:
+                mask = np.ma.nomask
+            elif mask_type == MaskTypes.ARRAY_FALSE:
+                mask = array_lib.full(data.shape, False)
+            else:
+                raise NotImplementedError
+
+            data_masked = array_lib.ma.masked_array(data, mask=mask)
+
+            if is_cube(array_or_cube):
+                array_or_cube.data = data_masked
+            else:
+                array_or_cube = data_masked
+
+        return array_or_cube
+
+    def _wrapper(*args, **kwargs):
+        # Will need to modify items in args, so can't be a tuple.
+        args = list(args)
+        mask_types_args = [get_mask_type(a) for a in args]
+        mask_types_kwargs = {k: get_mask_type(v) for k, v in kwargs.items()}
+
+        mask_types_all = mask_types_args + list(mask_types_kwargs.values())
+        mask_types_false = [
+            mt for mt in mask_types_all if mt in false_mask_types
+        ]
+
+        cube_store = CubeList()
+
+        masks_lifted = False
+        lift_count = len(mask_types_false)
+        if lift_count != 0:
+            uniform_false_mtype = mask_types_false[0]
+            if all([mt == uniform_false_mtype for mt in mask_types_false]):
+                # TODO: else raise a warning?
+
+                masks_lifted = True
+
+                for ix, arg in enumerate(args):
+                    if mask_types_args[ix] == uniform_false_mtype:
+                        args[ix] = lift_mask(arg, cube_store)
+                for key, kwarg in kwargs.items():
+                    if mask_types_kwargs[key] == uniform_false_mtype:
+                        kwargs[key] = lift_mask(kwarg, cube_store)
+
+        result = decorated_func(*args, **kwargs)
+
+        if masks_lifted:
+            for cube in cube_store:
```

Comment:
Is there a way to format this like a context manager or a try, except, finally statement so that cubes remain the same even if this fails partway through? I think this would only affect interactive sessions but it's probably still worth doing.

Reply:
Great idea, I have done that in the refactored version.